source/data-modeling/field-types.txt

```diff
@@ -0,0 +1,558 @@
+
+Phantom Custom Field Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can create a custom field type that saves a different value to the database
+than the value assigned in the application. This can be useful to have
+descriptive values in the application while storing more compact values in the
+database.
+
+The following example creates a ``ColorMapping`` type that uses the name of the
+color in the application, but stores the color as an integer in the database:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-phantom-field-type
+   :end-before: # end-phantom-field-type
+
+Dynamic Fields
+--------------
+
+You can instruct {+odm+} to create fields dynamically by setting the
+``Mongoid::Attributes::Dynamic`` module in your model. This allows {+odm+} to
+create fields based on an arbitrary hash, or by the documents stored in the
+database.
+
+The following example creates a ``Person`` class with dynamic fields:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-dynamic-field
+   :end-before: # end-dynamic-field
+
+You can also specify both fixed fields and dynamic fields within the same class.
+When doing so, {+odm+} treats all attributes for properties with field definitions
+according to their field type, and all other attributes as dynamic.
+
+When using dynamic fields in your application, you must initially set the value in one of the
+following ways:
+
+- Pass the attribute hash to the constructor.
+- Assign values by using the ``attributes=`` method.
+- Assign values by using the ``[]=`` method.
+- Assign values by using the ``write_attribute`` method.
+- Work with values that are already present in the database.
+
+If you don't initially set the value by using one of the preceding options,
+invoking the attribute returns a ``NoMethodError``.
+
+Reserved Characters
+-------------------
+
+Both {+odm+} and the MongoDB Query API use the ``.`` character to separate field
+names in nested documents, and use the ``$`` character at the beginning of a
+word to indicate a query operator. Because of this, you should avoid using these
+characters in your field names. 
+
+If your application requires the use of these characters, you can access the
+fields by calling the ``send`` method. The following example creates a ``User``
+class with fields that contain reserved characters. It then accesses the fields
+with the ``send`` method:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-reserved-characters
+   :end-before: # end-reserved-characters
+
+You can also access these fields by calling the ``read_attribute`` method.
+
+.. important::
+
+   Because updating and replacing fields containing these reserved characters
+   requires special operators, calling getters and setters on these fields
+   raises an ``InvalidDotDollarAssignment`` exception.
```

Comment:
S: add additional information section to link to other relevant guides (can leave as a todo)